x-pack/legacy/plugins/maps/public/components/map/feature_tooltip.js

```diff
@@ -7,136 +7,202 @@
 import React, { Fragment } from 'react';
 import {
   EuiButtonIcon,
-  EuiCallOut,
-  EuiLoadingSpinner,
-  EuiTextAlign,
+  EuiText,
+  EuiPagination,
+  EuiSelect,
+  EuiIconTip,
+  EuiHorizontalRule,
+  EuiFlexGroup,
+  EuiFlexItem
 } from '@elastic/eui';
+import _ from 'lodash';
 import { i18n } from '@kbn/i18n';
+import { FeatureProperties } from './feature_properties';
 
 
+const ALL_LAYERS = '_ALL_LAYERS_';
+const DEFAULT_PAGE_NUMBER = 0;
+
 export class FeatureTooltip extends React.Component {
 
   state = {
-    properties: undefined,
-    loadPropertiesErrorMsg: undefined,
+    uniqueLayers: [],
+    pageNumber: DEFAULT_PAGE_NUMBER,
+    layerIdFilter: ALL_LAYERS
   };
 
   componentDidMount() {
     this._isMounted = true;
-    this.prevLayerId = undefined;
-    this.prevFeatureId = undefined;
-    this._loadProperties();
   }
 
   componentDidUpdate() {
-    this._loadProperties();
+    this._loadUniqueLayers();
   }
 
   componentWillUnmount() {
     this._isMounted = false;
   }
 
-  _loadProperties = () => {
-    this._fetchProperties({
-      nextFeatureId: this.props.tooltipState.featureId,
-      nextLayerId: this.props.tooltipState.layerId,
-    });
-  }
+  _onLayerChange = (e) => {
 
-  _fetchProperties = async ({ nextLayerId, nextFeatureId }) => {
-    if (this.prevLayerId === nextLayerId && this.prevFeatureId === nextFeatureId) {
-      // do not reload same feature properties
+    const layerId = e.target.value;
+    if (this.state.layerIdFilter === layerId) {
       return;
     }
 
-    this.prevLayerId = nextLayerId;
-    this.prevFeatureId = nextFeatureId;
     this.setState({
-      properties: undefined,
-      loadPropertiesErrorMsg: undefined,
+      pageNumber: DEFAULT_PAGE_NUMBER,
+      layerIdFilter: layerId
     });
+  };
 
-    let properties;
-    try {
-      properties = await this.props.loadFeatureProperties({ layerId: nextLayerId, featureId: nextFeatureId });
-    } catch(error) {
-      if (this._isMounted) {
+  _onCloseTooltip = () => {
+    this.setState({
+      layerIdFilter: ALL_LAYERS,
+      pageNumber: DEFAULT_PAGE_NUMBER
+    }, () => {
+      this.props.closeTooltip();
+    });
+  };
+
+  _loadUniqueLayers = async () => {
+
+    const uniqueLayerIds = [];
```

Comment:
This is pretty heavy logic (especially if there are a lot of features under the curser) that will run every single state change. How about having `this.state.prevFeatures` and only generating the list if `this.state.prevFeatures !== this.props.features`?